man/tss2_verifyquote.1.md
% tss2_verifyquote(1) tpm2-tools | General Commands Manual
%
% APRIL 2019

# NAME

**tss2_verifyquote**(1) -

# SYNOPSIS

**tss2_verifyquote** [*OPTIONS*]

[common fapi references](common/tss2-fapi-references.md)

# DESCRIPTION

**tss2_verifyquote**(1) - This command verifies that the data returned by a quote is valid. This includes

  * Reconstructing the quoteInfo's PCR values from the eventLog (if an eventLog was provided)
  * Verifying the quoteInfo using the signature and the publicKeyPath

The used signature verification scheme is specified in the cryptographic profile (cf., **fapi-profile(5)**).

An application using tss2_verifyquote() will further have to

  * Assess the publicKey's trustworthiness
  * Assess the eventLog entries' trustworthiness

# OPTIONS

These are the available options:

  * **-Q**, **\--qualifyingData**=_FILENAME_ or _-_ (for stdin):

    A nonce provided by the caller to ensure freshness of the signature. Optional parameter.

  * **-l**, **\--pcrLog**=_FILENAME_ or _-_ (for stdin):

    Returns the PCR event log for the chosen PCR. Optional parameter.

    PCR event logs are a list (arbitrary length JSON array) of log entries with
    the following content.

        - recnum: Unique record number
        - pcr: PCR index
        - digest: The digests
        - type: The type of event. At the moment the only possible value is: "LINUX_IMA" (legacy IMA)
        - eventDigest: Digest of the event; e.g. the digest of the measured file
        - eventName: Name of the event; e.g. the name of the measured file.

  * **-q**, **\--quoteInfo**=_FILENAME_ or _-_ (for stdin):

    The JSON-encoded structure holding the inputs to the quote operation. This includes the digest value and PCR values.

  * **-k**, **\--publicKeyPath**=_STRING_:

    Identifies the signing key. MAY be a path to the public key hierarchy /ext.

  * **-i**, **\--signature**=_FILENAME_ or _-_ (for stdin):

    The signature over the quoted material.

[common tss2 options](common/tss2-options.md)

# EXAMPLE

```
    tss2_verifyquote --publicKeyPath="ext/myNewParent" --qualifyingData=qualifyingData.file --quoteInfo=quoteInfo.file --signature=signature.file --pcrLog=pcrLog.file
```

# RETURNS

0 on success or 1 on failure.

[footer](common/footer.md)


CHANGELOG
v2.2.2 - next:
*  fix make distcheck
v2.1.1 - next:
* Potential memory leak fix when tcti/sapi initialization fails.
* tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read
* listpcrs: remove one redundant call to tpm get cap
* listpcrs: fix for unsupported/disabled alg in -L
* build: use supported comment to suppress GCC7 fallthrough warning
* kdfa: allow to build with OpenSSL 1.1.x

v2.1.0 - 2017-7-24
* Fix readx and writex on multiple EINTR returns.
* Add support for the tabrmd TCTI. This is the new default.
* Change default socket port from 2323 (the old resourcemgr) to 2321 (default
  simulator port).
* Cherry-pick fix for CVE-2017-7524.
* Fix tpm2_listpcr command line option handling.
* Fix tpm2_getmanufec memory issues.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2.0-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/fapi/tss2_verifyquote.c
/* SPDX-License-Identifier: BSD-3-Clause */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "tools/fapi/tss2_template.h"

/* Context struct used to store passed command line parameters */
static struct cxt {
    char *publicKeyPath;
    char const *qualifyingData;
    char const *quoteInfo;
    char const *signature;
    char const *pcrLog;
} ctx;

/* Parse command line parameters */
static bool on_option(char key, char *value) {
    switch (key) {
    case 'Q':
        ctx.qualifyingData = value;
        break;
    case 'l':
        ctx.pcrLog = value;
        break;
    case 'q':
        ctx.quoteInfo = value;
        break;
    case 'k':
        ctx.publicKeyPath = value;
        break;
    case 'i':
        ctx.signature = value;
        break;
    }
    return true;
}

/* Define possible command line parameters */
static bool tss2_tool_onstart(tpm2_options **opts) {
    struct option topts[] = {
        {"publicKeyPath",   required_argument, NULL, 'k'},
        {"qualifyingData",  required_argument, NULL, 'Q'},
        {"quoteInfo",       required_argument, NULL, 'q'},
        {"signature",       required_argument, NULL, 'i'},
        {"pcrLog",          required_argument, NULL, 'l'}
    };
    return (*opts = tpm2_options_new ("k:Q:q:i:l:", ARRAY_LEN(topts), topts,
                                      on_option, NULL, 0)) != NULL;
}

/* Execute specific tool */
static int tss2_tool_onrun (FAPI_CONTEXT *fctx) {
    /* Check availability of required parameters */
    if (!ctx.quoteInfo) {
        fprintf (stderr, "quote info parameter not provided, use "\
            "--quoteInfo\n");
        return -1;
    }
    if (!ctx.publicKeyPath) {
        fprintf (stderr, "publicKeyPath parameter not provided, use "\
            "--publicKeyPath\n");
        return -1;
    }
    if (!ctx.signature) {
        fprintf (stderr, "signature parameter not provided, use"\
            " --signature\n");
        return -1;
    }

    /* Check exclusive access to stdin */
    int count_in = 0;
    if (ctx.qualifyingData && !strcmp (ctx.qualifyingData, "-")) count_in +=1;
    if (ctx.signature && !strcmp (ctx.signature, "-")) count_in +=1;
    if (ctx.quoteInfo && !strcmp (ctx.quoteInfo, "-")) count_in +=1;
    if (ctx.pcrLog && !strcmp (ctx.pcrLog, "-")) count_in +=1;
    if (count_in > 1) {
        fprintf (stderr, "Only one of --qualifyingData, --signature, "\
        " --quoteInfo and --pcrLog can read from - (standard input)\n");
        return -1;
    }

    /* Read qualifyingData, signature, quoteInfo and pcrLog from file */
    TSS2_RC r;
    uint8_t *qualifyingData = NULL;
    size_t qualifyingDataSize = 0;
    if (ctx.qualifyingData) {
        r = open_read_and_close (ctx.qualifyingData,
            (void**)&qualifyingData, &qualifyingDataSize);
        if (r) {
            return -1;
        }
    }

    uint8_t *signature = NULL;
    size_t signatureSize = 0;
    if (ctx.signature) {
        r = open_read_and_close (ctx.signature, (void**)&signature, &signatureSize);
        if (r) {
            free (qualifyingData);
            return -1;
        }
    }

    char *quoteInfo = NULL;
    if (ctx.quoteInfo) {
        r = open_read_and_close (ctx.quoteInfo, (void**)&quoteInfo, NULL);
        if (r) {
            free (qualifyingData);
            free (signature);
            return -1;
        }
    }

    char *pcrLog = NULL;
    if (ctx.pcrLog) {
        r = open_read_and_close (ctx.pcrLog, (void**)&pcrLog, NULL);
        if (r) {
            free (qualifyingData);
            free (signature);
            free (quoteInfo);
            return -1;
        }
    }

    /* Execute FAPI command with passed arguments */
    r = Fapi_VerifyQuote (fctx, ctx.publicKeyPath, qualifyingData,
        qualifyingDataSize, quoteInfo, signature, signatureSize,
        pcrLog);
    if (r != TSS2_RC_SUCCESS){
        free (qualifyingData);
        free (signature);
        free (quoteInfo);
        free (pcrLog);
        LOG_PERR ("Fapi_VerifyQuote", r);
        return 1;
    }

    free (qualifyingData);
    free (signature);
    free (quoteInfo);
    free (pcrLog);

    return 0;
}

TSS2_TOOL_REGISTER("verifyquote", tss2_tool_onstart, tss2_tool_onrun, NULL)
